packages/general/src/util/Semaphore.ts

```diff
@@ -0,0 +1,242 @@
+import { Duration } from "#time/Duration.js";
+import { Time, Timer } from "#time/Time.js";
+import { Instant } from "#time/TimeUnit.js";
+import { Logger } from "../log/Logger.js";
+import { Abort } from "./Abort.js";
+import { createPromise } from "./Promises.js";
+
+const logger = Logger.get("PromiseQueue");
+
+/**
+ * A work slot that must be released when work is complete.
+ */
+export interface WorkSlot extends Disposable {
+    /**
+     * Release the slot manually.
+     * This is called automatically when using `using` syntax.
+     */
+    release(): void;
+
+    /**
+     * Release the slot automatically when the object is garbage collected.
+     */
+    [Symbol.dispose](): void;
+}
+
+/**
+ * A queue that limits concurrent work using a slot-based approach.
+ *
+ * Instead of queueing promises or iterators directly, callers get a "work slot"
+ * which they hold while doing work. The slot must be released when work is complete.
+ */
+export class Semaphore {
+    readonly #delay: Duration;
+    readonly #queue = new Array<{
+        resolve: (slot: WorkSlot) => void;
+        reject: (reason?: any) => void;
+        abortListener?: () => void;
+        signal?: AbortSignal;
+    }>();
+    #delays = new Set<Timer>();
+    readonly #concurrency: number;
+    #runningCount = 0;
+    #delayCount = 0;
+    #closed = false;
+
+    constructor(concurrency = 1, delay = Instant) {
+        this.#concurrency = concurrency;
+        this.#delay = delay;
+    }
+
+    /**
+     * Get a work slot from the queue.
+     *
+     * This method returns a promise that resolves when a slot is available.
+     * The returned slot must be released when work is complete, either by
+     * calling `release()` or by using the `using` syntax.
+     *
+     * @param abort - Optional abort signal to cancel waiting for a slot
+     * @returns A disposable work slot
+     * @throws AbortedError if the abort signal is triggered before a slot is obtained
+     */
+    async obtainSlot(abort?: Abort.Signal): Promise<WorkSlot> {
+        if (this.#closed) {
+            throw new AbortedError("Queue is closed");
+        }
+
+        // Normalize signal
+        const signal = abort ? ("signal" in abort ? abort.signal : abort) : undefined;
+
+        // Check if already aborted
+        if (signal?.aborted) {
+            throw signal.reason ?? new AbortedError();
+        }
+
+        // If we have capacity, grant a slot immediately
+        if (this.#runningCount + this.#delayCount < this.#concurrency) {
+            return this.#grantSlot();
+        }
+
+        // Otherwise, queue up and wait
+        const { promise, resolver, rejecter } = createPromise<WorkSlot>();
+
+        const entry = {
+            resolve: resolver,
+            reject: rejecter,
+            signal,
+            abortListener: undefined as (() => void) | undefined,
+        };
+
+        // Set up abort listener if signal provided
+        if (signal) {
```

Comment:
If you have a local `abort` as described above then you can just convert the promise to `abort.race(promise)` and use a finally for cleanup.